File: 2221-check-if-a-parentheses-string-can-be-valid/check-if-a-parentheses-string-can-be-valid.js
/**
 * @param {string} s
 * @param {string} locked
 * @return {boolean}
 */
var canBeValid = function(s, locked) {
    // Odd length strings cannot form valid parentheses
    if (s.length % 2 !== 0) return false;
    
    // Forward pass
    let open = 0; // Tracks available '(' or unlocked indices
    for (let i = 0; i < s.length; i++) {
        if (locked[i] === '0' || s[i] === '(') {
            open++; // Treat as an opening parenthesis
        } else {
            open--; // Closing parenthesis needs a match
        }
        if (open < 0) return false; // Too many closing parentheses
    }
    
    // Backward pass
    let close = 0; // Tracks available ')' or unlocked indices
    for (let i = s.length - 1; i >= 0; i--) {
        if (locked[i] === '0' || s[i] === ')') {
            close++; // Treat as a closing parenthesis
        } else {
            close--; // Opening parenthesis needs a match
        }
        if (close < 0) return false; // Too many opening parentheses
    }

    return true; // Passed both forward and backward checks
};

File: 2221-check-if-a-parentheses-string-can-be-valid/check-if-a-parentheses-string-can-be-valid.py
class Solution(object):
    def canBeValid(self, s, locked):
        """
        :type s: str
        :type locked: str
        :rtype: bool
        """

        if len(s) %2 !=0:
            return False
        
        s_open = 0
        s_close = 0
        oc = 0

        # Left to right scan
        for i in range(len(s)):
            if locked[i] == '0':
                oc += 1
            else:
                if s[i] == '(':
                    s_open += 1
                else:
                    s_close += 1

            if s_close > s_open + oc:  # Too many ')'
                return False
        
        #Right to left 
        s_open = 0
        s_close = 0
        oc = 0
        
        for i in range(len(s) - 1, -1, -1):
            if locked[i] == '0':  
                oc += 1
            elif s[i] == ')':
                s_close += 1
            else:
                s_open += 1

            if s_open > s_close + oc:
                return False  

        return True

                
            

        